### ISSUE TITLE
ICS 20 should provide normative guidance on sending prefix information onwards in subsequent transfers

### ISSUE THREAD

**Issue by zmanian** [cosmos/ibc#232]
As currently specified ICS 20, the behavior specifies the source -> sink relations where a correct bank module will provide a prefix for all non-native assets.

I think it makes sense to prescribe normative behavior for source -> sink -> sink behavior (A->B->C) regarding extending the prefix chain.

A token transferred from A -> B -> C should be expected to be named `A/[port]/bank/denom` in thank bank module of a correct B and `B/[port]/bank/A/[port]/bank/denom` in the bank module of a correct C.

I'm actually having a little bit of difficulty parsing `prefix = "{packet/destPort}/{packet.destChannel}"`. 
I'm not sure if it means the opposite what I'm expecting in the above.

**Comment by cwgoes** [comment]
At present token denominations are tied to ports & channel identifiers (so necessarily unique) but *not* clients or connection identifiers (partly as a result of a discussion with @jaekwon).

This would mean that tokens transferred from A -> B -> C over two channels would have the prefix:

```
{portOnC}/{channelOnC}/{portOnB}/{channelOnB}
```

but that tokens transferred from A -> B -> C over one channel (in the future multi-hop world) would have:

```
{portOnC}/{channelOnC}
```

in both cases matching the redemption path.

This has the advantage of not tying tokens to a chain identifier which cannot be globally unique and thus might invite confusion, but has the disadvantage of possibly being less clear to the user.

Worthy of further discussion.

**Comment by cwgoes** [comment]
@mossid Did you have thoughts on this? I recall discussing but don't recall our conclusion.

**Comment by mossid** [comment]
In summary it was about removing portID from the token denoms, because we assume that the ocap security gives the token sending permission only to the modules that follow the rule of ICS20. So same as we don't put the moduleNames in the token minted internally(we are not saying `auth/atom` when the atoms are minted by the auth module), we should remove the portIDs from the interchain token denom format. Even the bank module upgrades, it will(and SHOULD) be backward compatible in sense of preserving the interchain balance tracking logic, so portID doesn't have to be specified per version either.

**Comment by cwgoes** [comment]
Hmm, I think we discussed this in person @mossid and I agreed with you, but now I am wondering again - channel identifiers are not globally unique, and I think we do want to differentiate where tokens came from (different modules, e.g. perhaps contracts, on the same chain) - so I don't know if we can omit the port, otherwise two modules might bind to different ports, use the same channel, and end up with the same denomination on the receiving chain, which would be problematic.

**Comment by fedekunze** [comment]
> This would mean that tokens transferred from A -> B -> C over two channels would have the prefix:
`{portOnC}/{channelOnC}/{portOnB}/{channelOnB}/denom`
but that tokens transferred from A -> B -> C over one channel (in the future multi-hop world) would have:
`{portOnC}/{channelOnC}/denom`
in both cases matching the redemption path.

A UX problem that I see now is that when you have token on a non-source chain (eg: `{portOnC}/{channelOnC}/{portOnB}/{channelOnB}`) and you want to send back the token though a direct connection C -> A to the source chain, the result will be:

`{portOnA}/{channelOnA}/{portOnC}/{channelOnC}/{portOnB}/{channelOnB}/denom` instead of just `denom` 

Sadly, since there's no token metadata https://github.com/cosmos/cosmos-sdk/issues/1980, can't deconstruct these prefix wrappers easily. 

Another concrete example, if there are multiple [bridges to ethereum](https://github.com/cosmos/peggy), we want to treat the Ether as the same, but since they were bridged through different zones, they are treated differently.

**Comment by cwgoes** [comment]
> A UX problem that I see now is that when you have token on a non-source chain (eg: `{portOnC}/{channelOnC}/{portOnB}/{channelOnB}`) and you want to send back the token though a direct connection C -> A to the source chain, the result will be: `{portOnA}/{channelOnA}/{portOnC}/{channelOnC}/{portOnB}/{channelOnB}/denom` instead of just denom

That's exactly correct - for now this is what we want though, sending back through a different path is not possible (this will require additional protocol logic to handle safety).

> Another concrete example, if there are multiple bridges to ethereum, we want to treat the Ether as the same, but since they were bridged through different zones, they are treated differently.

Do we? I think often we might not - e.g. they might have different finality thresholds / security levels.

**Comment by fedekunze** [comment]
> Do we? I think often we might not - e.g. they might have different finality thresholds / security levels.

Maybe at the protocol level we don't want to treat them equally. But from a wallet/user perspective, they are the same underlying token.

**Comment by fedekunze** [comment]
> sending back through a different path is not possible (this will require additional protocol logic to handle safety)

I don't think this is true. The current logic still allows you to transfer them through a direct connection, by escrowing the token in C and unescrowing the tokens that were originally transferred to B, on A.

**Comment by cwgoes** [comment]
> > Do we? I think often we might not - e.g. they might have different finality thresholds / security levels.
> 
> Maybe at the protocol level we don't want to treat them equally. But from a wallet/user perspective, they are the same underlying token.

I agree, I think this has to be handled by a higher layer though - e.g. perhaps there's a floating exchange rate between two pegs, one with more confirmations required & one with lower, and users can elect to pay a bit more for safety.

**Comment by cwgoes** [comment]
> > sending back through a different path is not possible (this will require additional protocol logic to handle safety)
> 
> I don't think this is true. The current logic still allows you to transfer them through a direct connection, by escrowing the token in C and unescrowing the tokens that were originally transferred to B, on A.
> 
> The problem is that the `denom` tokens will still be escrowed in B and C and they shouldn't be.

I don't think so - if you transfer from `A` to `B` to `C`, the denomination on `C` will be prefixed with first `B` then `A` (or the corresponding port/channels), so you won't be able to send it back to `A` directly. This is (for now) the intended behaviour.

**Comment by fedekunze** [comment]
> I agree, I think this has to be handled by a higher layer though 

Do you mean clients should do that?

**Comment by cwgoes** [comment]
> > I agree, I think this has to be handled by a higher layer though
> 
> Do you mean clients should do that?

Possibly intermediary protocols running "on top" of ICS 20, as well as wallets, certainly.

**Comment by cwgoes** [comment]
Closed by https://github.com/cosmos/ics/pull/375, which includes an explanation.

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
